### ISSUE TITLE
How does backprop work on a shared layer used twice in the same model?

### ISSUE THREAD

**Issue by Entropy4** [keras-team/keras#14388]
I made a custom layer for a learnable activation with 3 learnable parameters. In my model this layer is used at two different points. I wanted to know how keras calculates the gradients of these 3 parameters, since the gradients are different based on at which point you are calculating them.

I searched and googled for hours and I still couldn't find an answer. I'd be really grateful if someone would help clarify this doubt.

custom layer code:

```
class Diffact(keras.layers.Layer):
    def __init__(self):
        super(Diffact, self).__init__()
        self.k0 = self.add_weight(shape = (), initializer="random_normal", trainable=True)
        self.k1 = self.add_weight(shape = (), initializer="random_normal", trainable=True)
        self.k2 = self.add_weight(shape = (), initializer="random_normal", trainable=True)
        
    def call(self, inputs):
        return self.k0 + tf.multiply(inputs, self.k1) + tf.multiply(tf.math.multiply(inputs,inputs), self.k2)
```

model code:

```
inputs = keras.Input(shape=input_shape)
x = layers.Conv2D(32, kernel_size=(3, 3))(inputs)
diffact = Diffact()
x = diffact(x)
x = layers.MaxPooling2D(pool_size=(2, 2))(x)
x = layers.Conv2D(64, kernel_size=(3, 3))(x)
x = diffact(x)
x = layers.MaxPooling2D(pool_size=(2, 2))(x)
x = layers.Flatten()(x)
x = layers.Dropout(0.5)(x)
outputs = layers.Dense(num_classes, activation="softmax")(x)

model = keras.Model(inputs=inputs, outputs=outputs, name="mnist_diffact")
```

**Comment by ariG23498** [comment]
![image](https://user-images.githubusercontent.com/36856589/109801713-dea66800-7c44-11eb-9fc1-0d070b628ca8.png)

As far as I understand the architecture the backpropagation should go in this way:
- Dense
- Dropout
- Flatten
- MaxPooling
- Diffract: This is the first time this is going to get updated. Let's say it is updated to Diffract1.
- Conv2D
- MaxPooling
- Diffract: This is the second time this is going to get updated. The changes will be based on top of Diffract1.
- Conv2d

A simple experiment would be to manually calculate the gradients with `tf.gradients` and see what changes when. 😄

**Comment by sushreebarsa** [comment]
@Entropy4 Is this still an issue?
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. 
Please refer to this [thread](https://stackoverflow.com/questions/63222770/backpropagating-through-multiple-forward-passes) and let us know if it helps?
Thanks!
